jvm/src/main/scala/org/json4s/JValue.scala

```diff
@@ -0,0 +1,48 @@
+package org.json4s
+
+sealed trait JValue
+
+case object JNull extends JValue
+
+case class JString(value: String) extends JValue
+
+object JNumber{
+  private val mc = BigDecimal.defaultMathContext
+  @inline def apply(value: Int): JNumber = JNumber(BigDecimal(value))
+  @inline def apply(value: Byte): JNumber = JNumber(BigDecimal(value))
+  @inline def apply(value: Short): JNumber = JNumber(BigDecimal(value))
+  @inline def apply(value: Long): JNumber = JNumber(BigDecimal(value))
+  @inline def apply(value: BigInt): JNumber = JNumber(BigDecimal(value))
+  @inline def apply(value: Float): JNumber = JNumber({
+    // BigDecimal.decimal doesn't exist on 2.10, so this is just the 2.11 implementation
+    new BigDecimal(new java.math.BigDecimal(java.lang.Float.toString(value), mc), mc)
+  })
+  @inline def apply(value: Double): JNumber = JNumber(BigDecimal(value))
+}
+
+case class JNumber(value: BigDecimal) extends JValue {
+  @inline def to[B](implicit bigDecimalConverter: BigDecimalConverter[B]) = bigDecimalConverter(value)
+}
+
+sealed abstract class JBoolean extends JValue {
+  val value: Boolean
+}
+
+object JBoolean {
+  @inline def apply(x: Boolean): JBoolean = if (x) JTrue else JFalse
+  @inline def unapply(x: JBoolean): Option[Boolean] = Option(x.value)
```

Comment:
This could avoid a null check, which probably doesn't matter in the grand scheme, with `Some(x.value)`.